tokio-buf/src/buf_stream/mod.rs

```diff
@@ -0,0 +1,168 @@
+    fn poll_buf(&mut self) -> Poll<Option<Self::Item>, Self::Error>;
+
+    /// Returns the bounds on the remaining length of the stream.
+    ///
+    /// The size hint allows the caller to perform certain optimizations that
+    /// are dependent on the byte stream size. For example, `collect` uses the
+    /// size hint to pre-allocate enough capacity to store the entirety of the
+    /// data received from the byte stream.
+    ///
+    /// When `SizeHint::upper()` returns `Some` with a value equal to
+    /// `SizeHint::lower()`, this represents the exact number of bytes that will
+    /// be yielded by the `BufStream`.
+    ///
+    /// # Implementation notes
+    ///
+    /// While not enforced, implementations are expected to respect the values
+    /// returned from `SizeHint`. Any deviation is considered an implementation
+    /// bug. Consumers may rely on correctness in order to use the value as part
+    /// of protocol impelmentations. For example, an HTTP library may use the
+    /// size hint to set the `content-length` header.
+    ///
+    /// However, `size_hint` must not be trusted to omit bounds checks in unsafe
+    /// code. An incorrect implementation of `size_hint()` must not lead to
+    /// memory safety violations.
+    fn size_hint(&self) -> SizeHint {
+        SizeHint::default()
+    }
+
+    /// Indicates to the `BufStream` how much data the consumer is currently
+    /// able to process.
+    ///
+    /// The consume hint allows the stream to perform certain optimizations that
+    /// are dependent on the consumer's readiness. For example, the consume hint
+    /// may be used to request a remote peer to start sending up to `amount`
+    /// data.
+    ///
+    /// Calling `consume_hint` is not a requirement. If `consume_hint` is never
+    /// called, the stream should assume a default behavior. When `consume_hint`
+    /// is called, the stream should make a best effort to honor by the request.
+    ///
+    /// `amount` represents the number of bytes that the caller would like to
+    /// receive at the time the function is called. For example, if
+    /// `consume_hint` is called with 20, the consumer requests 20 bytes. The
+    /// stream may yield less than that. If the next call to `poll_buf` returns
+    /// 5 bytes, the consumer still has 15 bytes requested. At this point,
+    /// invoking `consume_hint` again with 20 resets the amount requested back
+    /// to 20 bytes.
+    ///
+    /// Calling `consume_hint` with 0 as the argument informs the stream that
+    /// the caller does not intend to call `poll_buf`. If `poll_buf` **is**
+    /// called, the stream may, but is not obligated to, return `NotReady` even
+    /// if it could produce data at that point. If it chooses to return
+    /// `NotReady`, when `consume_hint` is called with a non-zero argument, the
+    /// task must be notified in order to respect the `poll_buf` contract.
+    fn consume_hint(&mut self, amount: usize) {
```

Comment:
A possible alternative to this function is passing some hint as an argument to `poll_buf`. Like, `stream.poll_buf(1024 * 16)` or something...

Reply:
Is there an example of this providing value somewhere? I get the intention, but I'd personally prefer this left off until there's an implementation showing how it can be helpful.